l2geth/accounts/abi/abi.go

```diff
@@ -192,3 +194,25 @@ func (abi *ABI) EventByID(topic common.Hash) (*Event, error) {
 	}
 	return nil, fmt.Errorf("no event with id: %#x", topic.Hex())
 }
+
+// RevertSelector is a special function selector for revert reason unpacking.
+var RevertSelector = crypto.Keccak256([]byte("Error(string)"))[:4]
+
+// UnpackRevert resolves the abi-encoded revert reason. According to the solidity
+// spec https://solidity.readthedocs.io/en/latest/control-structures.html#revert,
+// the provided revert reason is abi-encoded as if it were a call to a function
+// `Error(string)`. So it's a special tool for it.
+func UnpackRevert(data []byte) (string, error) {
+	if len(data) < 4 {
+		return "", errors.New("invalid data for unpacking")
+	}
+	if !bytes.Equal(data[:4], RevertSelector) {
+		return "", errors.New("invalid data for unpacking")
+	}
+	typ, _ := NewType("string", "", nil)
+	unpacked, err := (Arguments{{Type: typ}}).UnpackValues(data[4:])
+	if err != nil {
+		return "", err
+	}
+	return unpacked[0].(string), nil
```

Comment:
Could this ever panic, e.g. when `len(unpacked) == 0`? Might be worth handling that case. Do we know that the cast will always work?

Reply:
I believe this will never panic. Unpacking errors should be caught above and because of the `RevertSelector` type we should be guaranteed to have a length=1 array.